Fix jetty 12.0.x transient timeouts

gregw opened the pull request:
Fixes #10234 and #10277 and also #10357. Builds on #10814 and #10801

lorban committed:
add reproducer

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add assertion to reproducer

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10357-fix-ServletChannel-ISE

gregw committed:
Do not fail callback in onFailure

gregw committed:
Transient timeouts. Fixes #10234 and #10277

gregw committed:
Transient timeouts. Fixes #10234 and #10277

gregw commented:
@lorban I'm making some progress on this, but there are lots of failures in servlet land because AsyncProducer has not been written with transient failures in mind.   It keeps reading again expecting to get the same chunk.   I have done some hacky fixes to improve this, but it is not enough.   Feel free to play in this branch and see if you can fix more.... but ultimately we probably want to take a step back a re-implement AsyncProducer with the new transient failure contract in mind.... but maybe bandaids can get us there.

gregw committed:
Transient timeouts. Fixes #10234 and #10277

lorban committed:
improve timeout-after-100-continue test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
introduce BlockingContentProducerTest and AsyncContentProducerTest

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix jetty-ee10-servlet pom

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
I reviewed the ee10 `ContentProducer` FSM, "ported" the tests from ee9 and added some tests for transient errors.

It looks like everything is in order now, and #10357 can be included in the list of issues this PR solves.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

olamy committed:
Jetty 12 merge from 11 and fix JAAS ldap login test (#10857)

* Merge from jetty-10.0.x to jetty-11.0.x

* Fixes #10812 - Correct awaitility dependency scope (#10813)

* Bump org.apache.logging.log4j:log4j-api from 2.20.0 to 2.21.1

Bumps org.apache.logging.log4j:log4j-api from 2.20.0 to 2.21.1.

---
updated-dependencies:
- dependency-name: org.apache.logging.log4j:log4j-api
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump maven.plugin-tools.version from 3.9.0 to 3.10.1

Bumps `maven.plugin-tools.version` from 3.9.0 to 3.10.1.

Updates `org.apache.maven.plugin-tools:maven-plugin-annotations` from 3.9.0 to 3.10.1
- [Release notes](https://github.com/apache/maven-plugin-tools/releases)
- [Commits](https://github.com/apache/maven-plugin-tools/compare/maven-plugin-tools-3.9.0...maven-plugin-tools-3.10.1)

Updates `org.apache.maven.plugin-tools:maven-plugin-tools-api` from 3.9.0 to 3.10.1
- [Release notes](https://github.com/apache/maven-plugin-tools/releases)
- [Commits](https://github.com/apache/maven-plugin-tools/compare/maven-plugin-tools-3.9.0...maven-plugin-tools-3.10.1)

---
updated-dependencies:
- dependency-name: org.apache.maven.plugin-tools:maven-plugin-annotations
  dependency-type: direct:production
  update-type: version-update:semver-minor
- dependency-name: org.apache.maven.plugin-tools:maven-plugin-tools-api
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump org.apache.maven.plugins:maven-clean-plugin from 3.3.1 to 3.3.2

Bumps [org.apache.maven.plugins:maven-clean-plugin](https://github.com/apache/maven-clean-plugin) from 3.3.1 to 3.3.2.
- [Release notes](https://github.com/apache/maven-clean-plugin/releases)
- [Commits](https://github.com/apache/maven-clean-plugin/compare/maven-clean-plugin-3.3.1...maven-clean-plugin-3.3.2)

---
updated-dependencies:
- dependency-name: org.apache.maven.plugins:maven-clean-plugin
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump org.apache.maven.plugins:maven-dependency-plugin

Bumps [org.apache.maven.plugins:maven-dependency-plugin](https://github.com/apache/maven-dependency-plugin) from 3.6.0 to 3.6.1.
- [Commits](https://github.com/apache/maven-dependency-plugin/compare/maven-dependency-plugin-3.6.0...maven-dependency-plugin-3.6.1)

---
updated-dependencies:
- dependency-name: org.apache.maven.plugins:maven-dependency-plugin
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump commons-io:commons-io from 2.14.0 to 2.15.0

Bumps commons-io:commons-io from 2.14.0 to 2.15.0.

---
updated-dependencies:
- dependency-name: commons-io:commons-io
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump org.apache.maven.plugins:maven-plugin-plugin from 3.9.0 to 3.10.1

Bumps [org.apache.maven.plugins:maven-plugin-plugin](https://github.com/apache/maven-plugin-tools) from 3.9.0 to 3.10.1.
- [Release notes](https://github.com/apache/maven-plugin-tools/releases)
- [Commits](https://github.com/apache/maven-plugin-tools/compare/maven-plugin-tools-3.9.0...maven-plugin-tools-3.10.1)

---
updated-dependencies:
- dependency-name: org.apache.maven.plugins:maven-plugin-plugin
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump maven.surefire.plugin.version from 3.1.2 to 3.2.1

Bumps `maven.surefire.plugin.version` from 3.1.2 to 3.2.1.

Updates `org.apache.maven.plugins:maven-failsafe-plugin` from 3.1.2 to 3.2.1
- [Release notes](https://github.com/apache/maven-surefire/releases)
- [Commits](https://github.com/apache/maven-surefire/compare/surefire-3.1.2...surefire-3.2.1)

Updates `org.apache.maven.plugins:maven-surefire-plugin` from 3.1.2 to 3.2.1
- [Release notes](https://github.com/apache/maven-surefire/releases)
- [Commits](https://github.com/apache/maven-surefire/compare/surefire-3.1.2...surefire-3.2.1)

Updates `org.apache.maven.surefire:surefire-junit47` from 3.1.2 to 3.2.1

---
updated-dependencies:
- dependency-name: org.apache.maven.plugins:maven-failsafe-plugin
  dependency-type: direct:production
  update-type: version-update:semver-minor
- dependency-name: org.apache.maven.plugins:maven-surefire-plugin
  dependency-type: direct:production
  update-type: version-update:semver-minor
- dependency-name: org.apache.maven.surefire:surefire-junit47
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump org.apache.maven.plugins:maven-checkstyle-plugin

Bumps [org.apache.maven.plugins:maven-checkstyle-plugin](https://github.com/apache/maven-checkstyle-plugin) from 3.3.0 to 3.3.1.
- [Commits](https://github.com/apache/maven-checkstyle-plugin/compare/maven-checkstyle-plugin-3.3.0...maven-checkstyle-plugin-3.3.1)

---
updated-dependencies:
- dependency-name: org.apache.maven.plugins:maven-checkstyle-plugin
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump apacheds.version from 2.0.0.AM26 to 2.0.0.AM27 (#10836)

* Bump apacheds.version from 2.0.0.AM26 to 2.0.0.AM27

Bumps `apacheds.version` from 2.0.0.AM26 to 2.0.0.AM27.

Updates `org.apache.directory.server:apacheds-test-framework` from 2.0.0.AM26 to 2.0.0.AM27
- [Commits](https://github.com/apache/directory-server/compare/2.0.0.AM26...2.0.0.AM27)

Updates `org.apache.directory.server:apacheds-server-integ` from 2.0.0.AM26 to 2.0.0.AM27
- [Commits](https://github.com/apache/directory-server/compare/2.0.0.AM26...2.0.0.AM27)

Updates `org.apache.directory.server:apacheds-core-integ` from 2.0.0.AM26 to 2.0.0.AM27
- [Commits](https://github.com/apache/directory-server/compare/2.0.0.AM26...2.0.0.AM27)

---
updated-dependencies:
- dependency-name: org.apache.directory.server:apacheds-test-framework
  dependency-type: direct:development
  update-type: version-update:semver-patch
- dependency-name: org.apache.directory.server:apacheds-server-integ
  dependency-type: direct:development
  update-type: version-update:semver-patch
- dependency-name: org.apache.directory.server:apacheds-core-integ
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* get this work

Signed-off-by: Olivier Lamy <olamy@apache.org>

---------

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Olivier Lamy <olamy@apache.org>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Olivier Lamy <olamy@apache.org>

* Bump org.infinispan:infinispan-bom from 11.0.17.Final to 11.0.18.Final (#10724)

Bumps [org.infinispan:infinispan-bom](https://github.com/infinispan/infinispan) from 11.0.17.Final to 11.0.18.Final.
- [Release notes](https://github.com/infinispan/infinispan/releases)
- [Changelog](https://github.com/infinispan/infinispan/blob/main/Jenkinsfile-release)
- [Commits](https://github.com/infinispan/infinispan/compare/11.0.17.Final...11.0.18.Final)

---
updated-dependencies:
- dependency-name: org.infinispan:infinispan-bom
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* not needed, bad merge

Signed-off-by: Olivier Lamy <olamy@apache.org>

* fix version

Signed-off-by: Olivier Lamy <olamy@apache.org>

---------

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Olivier Lamy <olamy@apache.org>
Co-authored-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>
Co-authored-by: Chad Wilson <chadw@thoughtworks.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

lorban commented:
We need to review all `Content.Chunk.isFailure(Chunk c)` calls as they currently all assume that a failure content is also a last content.

Now, there must be a distinction between terminal and transient failures, and in the latter case, one of the following actions has to be taken depending on the context:

 - Pass the chunk to another layer to decide what to do.
 - Ignore (silently?) the transient error.
 - Consider the transient failure as a terminal failure.

In the case of `Content.Source` read loops, if the transient error is ignored, we should loop again, while if it is considered terminal, the `Content.Source` must be explicitly ended with `fail()`.

There are a lot of places to review and many tests to add or modify to inject some transient errors and check the behavior.

lorban committed:
improve EE10 ContentProducer to more carefully assert transient and terminal errors + enable HttpInputIntegrationTest

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve HttpInputIntegrationTest + add HttpInputTransientErrorTest

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add give up on error tests in HttpInputTransientErrorTest

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add leak tracking

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
enable ee9 HttpInputIntegrationTest + add HttpInputTransientErrorTest

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
do not add connection: close to the response when the error is transient

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
@lorban perhaps we can hangout in your morning and go through this?   @sbordet join?

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

gregw committed:
Transient timeouts. Fixes #10234 and #10277

ongoing review

lorban committed:
add transient failure tests for all content sources

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
@lorban nice tests!

lorban committed:
add ResponseListeners and client Content tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
unrelated: fix ContentDecoder leak and retaining bugs + add tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
unrelated: restore ContentDecoder contract

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
unrelated: remove confusing comment

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
unrelated: update debug log

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add more transient failures tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

lorban committed:
unrelated: fix Chunk.asReadOnly() and add test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

lorban committed:
Merge branch 'fix/jetty-12.0.x/transient-timeouts' of github.com:jetty/jetty.project into fix/jetty-12.0.x/transient-timeouts

lorban committed:
Add MultiPartFormData.ContentSource tests + fixes

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Add AsyncContentListener test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Add GzipTransformer test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add javadoc and fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
use writable buffer

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@gregw @sbordet I reviewed all `Content.Chunk.isFailure(Chunk)` usages, added a bunch of tests, and fixed a few bugs, so this PR is now ready for review.

FYI only one of the changes we made in the first pass was wrong (it was in `MultiPart.AbstractContentSource.read()`, but there were other bugs).

Overall, transient failures work and can be elegantly retrofitted without changing the `Content.Source` contract.

lorban committed:
fix chunks duplication

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
I can't review this PR, so over to you @sbordet

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/Response.java:
```diff
@@ -190,6 +190,8 @@ default void onContentSource(Response response, Content.Source contentSource)
             {
                 response.abort(chunk.getFailure());
+                if (!chunk.isLast())
+                    contentSource.fail(chunk.getFailure());
```
This block is ok. But should we fail in the `catch` below, or just `LOG.info()` like we do when notifying user code?

sbordet replied:
@lorban I mean this:

```java
            try
            {
                onContent(response, chunk, () -> contentSource.demand(demandCallback));
                chunk.release();
            }
            catch (Throwable x)
            {
                chunk.release();
                response.abort(x);
                **contentSource.fail(x);**
            }
```

lorban replied:
You're right, but this is a pre-existing bug. Let me fix it.

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpReceiver.java:
```diff
@@ -598,34 +603,47 @@ protected Content.Chunk transform(Content.Chunk inputChunk)
+                if (LOG.isDebugEnabled())
+                    LOG.debug("decoding produced no content");
+
+                if (!_chunk.hasRemaining())
+                {
+                    Content.Chunk result = _chunk.isLast() ? Content.Chunk.EOF : null;
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Could not decode more from this chunk, releasing it, r={}", result);
+                    _chunk.release();
+                    _chunk = null;
+                    return result;
+                }
```
Please review this logic against the original.
Contract is that if decoded == null then need more input, otherwise trust the decoded.

lorban replied:
You're right, let me rework this to minimize the changes.

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/AsyncContentListenerTest.java:
```diff
@@ -0,0 +1,65 @@
+        List<Content.Chunk> collectedChunks = new ArrayList<>();
+        Response.AsyncContentListener asyncContentListener = (response, chunk, demander) ->
+        {
+            collectedChunks.add(chunk);
```
Technically you should retain the chunk here, because it's stored for later use.
For this test it is not necessary only because the buffers are not consumed, nor recycled, but would be best if we can stick to the doctrine :)

lorban replied:
Let me fix this.

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/TestSource.java:
```diff
@@ -0,0 +1,124 @@
+
+import org.eclipse.jetty.io.Content;
+
+public class TestSource implements Content.Source
```
Can you reuse `ChunksContentSource` instead of writing this class from scratch?
I'm fine to have a `TestSource` class, just wondering if we can reuse `ChunksContentSource` for its implementation.

The constructor of this class breaks the doctrine, so this class needs to have javadocs that say that's special, and can only receive newly created `Chunk`s, etc.

Would be great to avoid all these special cases, otherwise we may possibly introduce bugs because this class does not behave as per doctrine, but the tests pass because they make assumptions.

lorban replied:
`TestSource` can inject null reads, which I do not think we want in any production-ready content source impl. But you're right that it should respect the doctrine, even if it's a test class.

Let me fix this.

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/transport/ResponseListenersTest.java:
```diff
@@ -85,65 +86,127 @@ public void run()
+                        source.demand(this);
+                        return;
+                    }
+                    chunk.release();
```
This release should not be done, as the chunk is stored for later use, as per doctrine.

lorban replied:
You're right, let me fix this.

sbordet commented on jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/transport/ResponseListenersTest.java:
```diff
@@ -85,65 +86,127 @@ public void run()
+        assertThat(chunks.get(3).isLast(), is(false));
+        assertThat(Content.Chunk.isFailure(chunks.get(4), false), is(true));
+        assertThat(chunks.get(4).getFailure(), instanceOf(TimeoutException.class));
+        assertThat(Content.Chunk.isFailure(chunks.get(5), true), is(true));
+        assertThat(chunks.get(5).getFailure(), instanceOf(NumberFormatException.class));
```
This is a bit unexpected. I would have expected (4 -> transient failure), (5 -> transient failure), (6 -> terminal failure), (7 -> terminal failure).

Is it a glitch of the test?

lorban replied:
No, it's an actual glitch in the implementation as failing a multiplexed content source mistakenly replaces its `ALREADY_READ_CHUNK` with the failure chunk. Let me fix this.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPart.java:
```diff
@@ -407,7 +407,13 @@ public Content.Source newContentSource()
+                    .map(chunk ->
+                    {
+                        // TODO is this a reliable way to duplicate a chunk while eventually severing it from its retainable?
+                        if (Content.Chunk.isFailure(chunk))
+                            return chunk;
```
Move this to a `filter()` call.

lorban replied:
Once the TODO below is resolved, yes, I can do that.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPart.java:
```diff
@@ -407,7 +407,13 @@ public Content.Source newContentSource()
-                    .map(chunk -> Content.Chunk.from(chunk.getByteBuffer().slice(), chunk.isLast()))
+                    .map(chunk ->
+                    {
+                        // TODO is this a reliable way to duplicate a chunk while eventually severing it from its retainable?
```
Remove TODO

lorban replied:
Did we give up on the idea of trailer chunks, and other kind of non-content, non-failure, non-EOF, non-EMPTY chunks?

If we did, this comment can go away, otherwise this code won't correctly duplicate the chunk.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MultiPart.java:
```diff
@@ -759,8 +765,13 @@ public Content.Chunk read()
+                        yield null;
+                    if (Content.Chunk.isFailure(chunk, true))
+                    {
+                        errorChunk = chunk;
```
`errorChunk` should be set holding the lock.

lorban replied:
Good catch, but the fact that the lock only protects a subset of the fields (`errorChunk` but not `part` nor `state`) is confusing and should be documented. But that's a job for another PR.

Let me add the lock.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -95,6 +95,11 @@ public static void copy(Source source, Sink sink, Chunk.Processor chunkProcessor
      * <p>A source of content that can be read with a read/demand model.</p>
+     * <p>To avoid leaking its resources, a source <b>must</b> either:</p>
+     * <ul>
+     * <li>be read until it returns the {@link Chunk#isLast() last chunk}</li>
```
```suggestion
     * <li>be read until it returns a {@link Chunk#isLast() last chunk}, either EOF or a terminal failure</li>
```

lorban replied:
Done.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourceCompletableFuture.java:
```diff
@@ -83,9 +85,17 @@ public void parse()
+                {
+                    if (onTransientFailure(chunk.getFailure()))
+                        continue;
+                    completeExceptionally(chunk.getFailure());
+                    _content.fail(chunk.getFailure());
```
These 2 lines should be swapped, as we want to notify the CF last, because it can act before we failed `_content`.

lorban replied:
Done.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java:
```diff
@@ -129,6 +129,8 @@ private void process()
                 {
                     terminate();
                     subscriber.onError(chunk.getFailure());
+                    if (!chunk.isLast())
+                        content.fail(chunk.getFailure());
```
I'd swap these lines and notify the `subscriber` as last, as it can act before the content is failed.

lorban replied:
Done.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java:
```diff
@@ -61,14 +61,7 @@ protected Action process() throws Throwable
-            succeeded();
-            return Action.SCHEDULED;
-        }
+            throw current.getFailure();
```
If the read a transient failure, I think we should behave as if we read `null` for the goal of copying.

lorban replied:
I disagree as that would silently ignore idle timeouts.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceByteBuffer.java:
```diff
@@ -47,6 +47,8 @@ public void run()
             if (Content.Chunk.isFailure(chunk))
             {
                 promise.failed(chunk.getFailure());
+                if (!chunk.isLast())
+                    source.fail(chunk.getFailure());
```
This is reading and accumulating into a single BB.
If we have a transient failure, should we not just continue as if we read `null`?

lorban replied:
That would silently ignore idle timeouts, which I think isn't a reasonable default. We could eventually add a mechanism that allows taking decisions on what to do with failures, but that's outside the scope of this PR.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceString.java:
```diff
@@ -45,6 +45,8 @@ public void convert()
             if (Content.Chunk.isFailure(chunk))
             {
                 promise.failed(chunk.getFailure());
+                if (!chunk.isLast())
+                    content.fail(chunk.getFailure());
```
Ditto.

lorban replied:
Same answer.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceConsumer.java:
```diff
@@ -44,6 +44,8 @@ public void run()
             if (Content.Chunk.isFailure(chunk))
             {
                 callback.failed(chunk.getFailure());
+                if (!chunk.isLast())
+                    source.fail(chunk.getFailure());
```
Ditto.

lorban replied:
Same answer.

sbordet commented on jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/TestSource.java:
```diff
@@ -0,0 +1,109 @@
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class TestSource implements Content.Source
```
Same comment as for the duplicate in `jetty-client`.

lorban replied:
Not sure what do do here.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -432,34 +422,20 @@ else if (ExceptionUtil.areNotAssociated(_failure.getFailure(), x) && _failure.ge
-                        onFailure = _onFailure;
-                    }
 
+                Consumer<Throwable> onFailure = _onFailure;
```
Field `_onFailure` should be read with the lock held.
Also, not sure we want to read it here, it's too early. We only want it when we execute its runnable (and previous runnables may generate errors).

lorban replied:
AFAICT, this code is within the context of a lock.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -432,34 +422,20 @@ else if (ExceptionUtil.areNotAssociated(_failure.getFailure(), x) && _failure.ge
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("failing callback in {}", this, x);
-                        request._callback.failed(x);
-                    }
```
Is it correct that this block has been removed?

lorban replied:
Yes, it's part of the change that `testAsyncReadHttpIdleTimeoutOverridesIdleTimeout` covers, see the big TODO that was removed from that test.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -1094,7 +1094,17 @@ public void earlyEOF()
                 if (Content.Chunk.isFailure(stream._chunk))
-                    stream._chunk.getFailure().addSuppressed(bad);
+                {
+                    if (stream._chunk.isLast())
```
Assign `stream._chunk` to a local variable for better readability.

lorban replied:
Done.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/TestSource.java:
```diff
@@ -0,0 +1,111 @@
+
+import org.eclipse.jetty.io.Content;
+
+public class TestSource implements Content.Source
```
Gosh, 3rd time :(
Can't we move these tests to a common place?

lorban replied:
Unfortunately, we don't want this class anywhere in the production code, we don't want to productize it and there is no easy way to share a test util class...

sbordet commented on jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/ServerTimeoutsTest.java:
```diff
@@ -75,7 +76,10 @@ public boolean handle(Request request, Response response, Callback callback)
             {
 
                 if (listener)
+                {
```
Parameter `listener` can be either `true` or `false`, but this test is annotated with `@MethodSource("transportsAndTrueIdleTimeoutListeners")` which seems to indicate only `true`.
If the method-source name correct?

lorban replied:
`transportsAndTrueIdleTimeoutListeners` does provide both `true` and `false` as `listener` values but its name is indeed confusing.

Let me rename it.

sbordet commented on jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/ServerTimeoutsTest.java:
```diff
@@ -130,15 +134,13 @@ public boolean handle(Request request, Response response, Callback callback)
         // Can read again
         assertNull(requestRef.get().read());
-        */
+
```
Remove empty line.

lorban replied:
Done.

sbordet commented on jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/ServerTimeoutsTest.java:
```diff
@@ -198,6 +200,7 @@ public void testIdleTimeoutErrorListenerReturnsFalseThenTrue(Transport transport
             public boolean handle(Request request, Response response, Callback callback)
             {
                 request.addIdleTimeoutListener(t -> error.getAndSet(t) != null);
+                request.addFailureListener(callback::failed);
```
Please fix the comment before the asserts few lines below.
The first time the idle timeout listener returns `false`, not `true`.

lorban replied:
Done.

sbordet commented on jetty-ee10/jetty-ee10-proxy/src/main/java/org/eclipse/jetty/ee10/proxy/ProxyServlet.java:
```diff
@@ -260,6 +260,8 @@ public Content.Chunk read()
             if (Content.Chunk.isFailure(chunk))
             {
                 onClientRequestFailure(request, proxyRequest, response, chunk.getFailure());
+                if (!chunk.isLast())
+                    super.fail(chunk.getFailure(), true);
```
I'd swap the lines.

lorban replied:
Done.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducer.java:
```diff
@@ -292,10 +304,9 @@ private Content.Chunk produceChunk()
-                    _servletChannel.getServletRequestState().onContentAdded();
-                }
+                _servletChannel.getServletRequestState().onContentAdded();
+                if (Content.Chunk.isFailure(_chunk, false))
+                    return _chunk;
```
Would be great if we can fold this logic into the existing one above.
Can we call `continue;` here to loop around and execute the logic at the top?
This would get rid of the duplicated logic below this code.

gregw replied:
I think we can do the loop as @sbordet suggested.    However, when I tried it, I thought we should also be calling `Chunk.next(_chunk)` on the transient failure, but that breaks tests.     I thought that was our general pattern, so we need to comment why it does not apply here.     My code was:

```java
    private Content.Chunk produceChunk()
    {
        if (LOG.isDebugEnabled())
            LOG.debug("produceChunk() {}", this);

        while (true)
        {
            if (_chunk != null)
            {
                if (Content.Chunk.isFailure(_chunk, false))
                {
                    // TODO comment why these lines are not needed here:
                    // Content.Chunk chunk = _chunk;
                    // _chunk = Content.Chunk.next(chunk);
                    return _chunk;
                }
                if (_chunk.isLast() || _chunk.hasRemaining())
                {
                    if (LOG.isDebugEnabled())
                        LOG.debug("chunk not yet depleted, returning it {}", this);
                    return _chunk;
                }
                if (LOG.isDebugEnabled())
                    LOG.debug("current chunk depleted {}", this);
                _chunk.release();
                _chunk = null;
            }
            else
            {
                if (LOG.isDebugEnabled())
                    LOG.debug("reading new chunk {}", this);
                _chunk = readChunk();
                if (_chunk == null)
                {
                    if (LOG.isDebugEnabled())
                        LOG.debug("channel has no new chunk {}", this);
                    return null;
                }
                _servletChannel.getServletRequestState().onContentAdded();
            }
        }
    }
```

gregw replied:
Ah - I have answered my own question:  `produceChunk()` may be called by `available()` and `isReady()`, which do not consume the chunk, so we cannot move to the next chunk here. Only in `nextChunk()` should we move to the next chunk.   I'll commit my code changes.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/BlockingContentProducer.java:
```diff
@@ -149,19 +149,18 @@ public boolean onContentProducible()
+        // the dispatched thread.
         // Calling _asyncContentProducer.onContentProducible() changes the channel state from WAITING to WOKEN which
-        // would prevent the subsequent call to HttpChannelState.asyncError() from rescheduling the thread.
-        // AsyncServletTest.testStartAsyncThenClientStreamIdleTimeout() tests this.
+        // would prevent the async error thread from noticing that a redispatching is needed.
```
This paragraph seems to be outdated? We do not call `_asyncContentProducer.onContentProducible()` below.

lorban replied:
It's what the comment is about: calling `_asyncContentProducer.onContentProducible()` is wrong as the dispatching logic is very different between blocking and async.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/BlockingContentProducer.java:
```diff
@@ -149,19 +149,18 @@ public boolean onContentProducible()
-            // Call nextChunk() to switch the input state back to IDLE, otherwise we would stay UNREADY.
-            _asyncContentProducer.nextChunk();
+            // Switch the input state back to IDLE, otherwise we would stay UNREADY.
+            _asyncContentProducer.getServletChannel().getServletRequestState().onReadIdle();
```
Is this ok? Seems weird because the async version calls `onReadReady()` which seems more appropriate for this event.
If it is ok, needs additional comments as to why we go idle rather than ready.

lorban replied:
The `READY` input state is meaningless in blocking mode: it means data has arrived and we may have to redispatch the request. But in blocking mode, we know that if are in input state `UNREADY`, the request is blocked on the semaphore; there is no need to redispatch anything as releasing the semaphore is all you need to unblock the request.

This is what the two comment blocks in this method try to explain.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/HttpInput.java:
```diff
@@ -42,8 +41,8 @@ public class HttpInput extends ServletInputStream implements Runnable
     private final byte[] _oneByteBuffer = new byte[1];
-    private final BlockingContentProducer _blockingContentProducer;
-    private final AsyncContentProducer _asyncContentProducer;
+    final BlockingContentProducer _blockingContentProducer;
+    final AsyncContentProducer _asyncContentProducer;
```
Not a big fan, see below.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/AsyncContentProducerTest.java:
```diff
@@ -0,0 +1,208 @@
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class AsyncContentProducerTest extends AbstractContentProducerTest
```
Not a big fan of these tests, as they require a framework to run (base test class + mocks), a 9 parameters method, and only test some parts (the content producers) but not others (httpinput).

Can we rewrite these in a simpler way, as they don't seem to test anything specific just some normal content cases (e.g. content no eof, content with eof, content with failure, etc.).

gregw replied:
On the contrary, I am a big fan of these tests.   We have imagined a `ContentProducer` contract and these tests are checking that contract.   It may well be that `HttpInput` uses the `ContentProducers` differently, but then that is either an issue with the contract not being good enough or that `HttpInput` is not respecting the contract.

I think having these simple tests is definitely worth it.   But it does not mean we should not also test with `HttpInput`

lorban replied:
`ContentProducer`s are not as testable independently as I wish they should be, but those tests have assertions that would be very complicated to replicate with integration tests.

I think they're a reasonable compromise.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/BlockingContentProducerTest.java:
```diff
@@ -0,0 +1,184 @@
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class BlockingContentProducerTest extends AbstractContentProducerTest
```
Ditto

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/MockConnectionMetaData.java:
```diff
@@ -0,0 +1,132 @@
+import org.eclipse.jetty.util.HostPort;
+
+// TODO shared copy of this class
+public class MockConnectionMetaData extends Attributes.Mapped implements ConnectionMetaData
```
Let's try to not use mocks.

gregw replied:
I know mocks are not a good universal test strategy and sometimes they are used way too much and take too much effort and often test the mock rather than the class....
But sometimes there is vital code that you really want to explicitly test and there is now other way to create a unit test to check it.    By all means create an integration test as well to ensure the mock test is not a bad test, but I see no reason to not have some reasonable mock tests when appropriate.

I think we should be mock reluctant, but not mock-phobic.    If there is a way to do a unit test without a mock, let's do that.    But I don't think the solution should be to always turn a unit test into an integration test.   There is value it checking the contract of a component is implemented as we imagined.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/MockRequest.java:
```diff
@@ -0,0 +1,173 @@
+import org.eclipse.jetty.server.Session;
+import org.eclipse.jetty.server.TunnelSupport;
+
+public class MockRequest implements Request
```
Ditto.

sbordet commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/AbstractTest.java:
```diff
@@ -143,7 +146,16 @@ protected void prepareServer(Transport transport, HttpServlet servlet) throws Ex
     protected Server newServer()
     {
-        QueuedThreadPool serverThreads = new QueuedThreadPool();
+        jobFailures = new CopyOnWriteArrayList<>();
+        QueuedThreadPool serverThreads = new QueuedThreadPool()
+        {
+            @Override
+            protected void onJobFailure(Throwable x)
+            {
+                super.onJobFailure(x);
+                jobFailures.add(x);
+            }
+        };
```
This is so weird and only used in 1 test at the Servlet level, and not sure why we are checking the failures of thread pool jobs?

lorban replied:
It's about https://github.com/jetty/jetty.project/issues/10357#issuecomment-1786824936, see below.

sbordet commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/HttpClientContinueTest.java:
```diff
@@ -676,6 +678,56 @@ public void onComplete(Result result)
+
+        assertTrue(serverLatch.await(5, TimeUnit.SECONDS));
+        assertTrue(clientLatch.await(5, TimeUnit.SECONDS));
+        assertThat(jobFailures, empty());
```
Why this last assert? Seems way too far away in abstractions layers to be tested here.

lorban replied:
This was the only half-sane way I found to create a reliable reproducer for the problem initially reported in #10357. I added this test as part of PR #10814 upon which this PR is based.

I detailed the actual problem here: https://github.com/jetty/jetty.project/issues/10357#issuecomment-1786824936 and discussed with @gregw here: https://github.com/jetty/jetty.project/pull/10801#discussion_r1376991169

It's not directly related to the transient failures, but it's part of the family of problems for which transient failures are a piece of the puzzle.

lorban replied:
After discussing this, we agreed to remove the assertion.

sbordet commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/ServerTimeoutsTest.java:
```diff
@@ -463,7 +462,8 @@ protected void service(HttpServletRequest request, HttpServletResponse response)
                     public void onDataAvailable() throws IOException
                     {
-                        assertEquals(0, input.read());
+                        while (input.isReady())
+                            assertEquals(0, input.read());
```
Why this? The client only sends 1 byte, we have been called in `onDataAvailable()` so when we read we must read 1 byte, as the previous code was doing.

gregw replied:
I think the idea was to check that both an `isReady` and a `read` could be done.    So it probably could have been done with an `if` rather than a `while`, but the `while` is the more natural style.

However, we should probably also have a test of a `read` without a prior `isReady` call (as it was before... we should test both).

lorban replied:
Let me add such test.

sbordet commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-client-transports/src/test/java/org/eclipse/jetty/ee10/test/client/transport/ServerTimeoutsTest.java:
```diff
@@ -648,19 +641,13 @@ public BlockingReadServlet(CountDownLatch handlerLatch)
-                handlerLatch.countDown();
-                throw x;
-            }
+            IOException x = assertThrows(IOException.class, input::read);
+            handlerLatch.countDown();
+            throw x;
```
Nice!

sbordet commented on jetty-ee9/jetty-ee9-proxy/src/main/java/org/eclipse/jetty/ee9/proxy/ProxyServlet.java:
```diff
@@ -260,6 +260,8 @@ public Content.Chunk read()
             if (Content.Chunk.isFailure(chunk))
             {
                 onClientRequestFailure(request, proxyRequest, response, chunk.getFailure());
+                if (!chunk.isLast())
+                    super.fail(chunk.getFailure());
```
Swap lines.

lorban replied:
Done.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

gregw committed:
simplified produceChunk

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
revert to previous behavior which was correct, and document it in the test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve comment

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
rework doc to account for transient failures

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpReceiver.java:
```diff
@@ -602,6 +607,12 @@ protected Content.Chunk transform(Content.Chunk inputChunk)
                 {
                     // The decoded ByteBuffer is a transformed "copy" of the
                     // compressed one, so it has its own reference counter.
+                    if (!decodedBuffer.canRetain())
```
Can you invert the `if` so it's more readable.

lorban replied:
Done.

sbordet commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/ResponseListeners.java:
```diff
@@ -679,10 +680,18 @@ public void fail(Throwable failure)
+                    if (currentChunk != null && currentChunk != ALREADY_READ_CHUNK)
+                        currentChunk.release();
+                    this.chunk = Content.Chunk.from(failure);
+                }
                 onDemandCallback();
                 registerFailure(this, failure);
```
I think the 2 calls should be swapped.
If not, needs a comment why this order is the way it is.

lorban replied:
Done.

lorban replied:
Ok, swapped them.

sbordet commented on documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/ContentDocs.java:
```diff
@@ -54,13 +54,33 @@ public void read(Content.Source source)
                 }
 
+                // If there is a transient failure reading, handle it.
+                if (Content.Chunk.isFailure(chunk, false))
```
I prefer a single `if` for the failure case, and explicit variables, e.g. `var transientFailure = !chunk.isLast()`.

lorban replied:
Done.

lorban committed:
rework ChunksContentSource to support null chunks, and simplify TestSource

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
simplify the failure block

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ChunksContentSource.java:
```diff
@@ -60,18 +75,16 @@ public long getLength()
             chunk = iterator.next();
-            last = !iterator.hasNext();
-            if (last)
+            if (chunk != null && chunk.isLast())
+                terminated = chunk;
+            if (terminated == null && !iterator.hasNext())
```
Don't want to remember as `terminated` a "normal" chunk that has bytes _and_ is last.
Can we use here `Chunk.next()`?

lorban replied:
Yes, this should be:
```java
terminated = Content.Chunk.next(chunk);
```
Done.

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet committed:
Improved documentation about `Content.Source`.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12.0.x/transient-timeouts

lorban committed:
Merge branch 'fix/jetty-12.0.x/transient-timeouts' of github.com:jetty/jetty.project into fix/jetty-12.0.x/transient-timeouts

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>
